src/sotauptaneclient.cc

```diff
@@ -249,3 +236,34 @@ void SotaUptaneClient::runForever(command::Channel *commands_channel) {
     }
   }
 }
+
+// TODO: the function can't currently return any errors. The problem of error reporting from secondaries should be
+// solved on a system (backend+frontend) error.
+// TODO: the function blocks until it updates all the secondaries. Consider non-blocking operation.
+void SotaUptaneClient::updateSecondaries(std::vector<Uptane::Target> targets) {
+  std::vector<Uptane::Target>::const_iterator it;
+
+  Uptane::MetaPack meta;
+  if (!uptane_repo.currentMeta(&meta)) {
+    // couldn't get current metadata package, can't proceed
+    return;
+  }
+  // target images should already have been downloaded to metadata_path/targets/
+  for (it = targets.begin(); it != targets.end(); ++it) {
+    std::map<std::string, boost::shared_ptr<Uptane::SecondaryInterface> >::iterator sec =
+        secondaries.find(it->ecu_identifier());
+    if (sec != secondaries.end()) {
+      std::string firmware_path = uptane_repo.image.getTargetPath(*it);
+      if (!boost::filesystem::exists(firmware_path)) continue;
+
+      if (!sec->second->putMetadata(meta)) {
+        // connection error while putting metadata, can't upload firmware
+        continue;
+      }
+
+      std::string firmware = Utils::readFile(firmware_path);
+
+      sec->second->sendFirmware((const uint8_t *)firmware.c_str(), firmware.size());
```

Comment:
What's the benefit of using `uint8_t`? Currently (for `ManagedSecondary`), we're going file -> `string` -> C-string -> `uint8_t` -> C-string -> `string` -> file. I know that we may want to rethink that anyway because of large file sizes, but I think we can do without some of these layers of casts. Or we should just read into a binary stream with something like `fopen`.

Reply:
Yes, maybe the interface should be just `sendFirmware(std::string data)`